packages/block-editor/src/components/block-context/README.md

```diff
@@ -0,0 +1,57 @@
+Block Context
+=============
+
+Block Context is a React implementation of WordPress's block context. Block context, much like [React's context](https://reactjs.org/docs/context.html), is a method for passing and inheriting values deeply through a hierarchy of blocks. Because of the similarities with React's context, the client-side implementation here is quite minimal. It is complemented by equivalent behaviors in the server-side rendering of a block.
+
+Note that the implementation of Block Context is distinct from [the `BlockEdit` context](../block-edit). While it is true that both provide context relevant for the editing of a block, Block Context is implemented separately so as to prioritize it as most identifiable amongst the machinery of block context, and not amongst other client-side editing context of a block.
+
+## Usage
+
+Currently, only the [Provider component](https://reactjs.org/docs/context.html#contextprovider) is made availble on the public interface of the `@wordpress/block-editor` module. This can be used to add or override context which can then be consumed by blocks rendered within that context.
+
+```js
+import { BlockContextProvider } from '@wordpress/block-editor';
+
+function MyCustomPostEditor() {
+	return (
+		<BlockContextProvider value={ { postId: 1, postType: 'post' } }>
```

Comment:
I'm not fully convinced that the proposed name is clear enough. I took me some time to understand the difference between the `BlockContext` and `BlockEditContext`. The fact that it can wrap `BlockEditorProvider` made me think that it could be called `BlockEditorContext` to make it explicit that it isn't scoped to the individual block as `BlockEdit` context (in this case all values are always overridden in the child blocks).

Reply:
> it could be called `BlockEditorContext` to make it explicit that it isn't scoped to the individual block as `BlockEdit` context

My first reaction to this is that this naming would be confusing as well, because the context has nothing to do with the block editor itself. It is about the block, though granted not at an individual level, but rather as an aggregate of blocks which provide context.

I see your point though. I'll want to give some more time to think about this.

Reply:
Thinking about this more, I'm feeling inclined to keep the current name. Or at least, my thinking is:

- This is implementing the thing we call "Block Context". A provider which creates this context value could thus reasonably be called a "Block Context Provider".
- Everything in this module is going to have _something_ to do with editing. Including "Editor" or "Edit" qualifiers doesn't seem to me to add much value for this specific usage.

As to relationships with other providers in this module:

- [`useBlockEditContext` ("block edit context")](https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/block-edit/context.js) seems named reasonably appropriate as well, given that its usage is largely relevant for `edit` implementations of a block.
- [`BlockListBlock`'s `BlockContext`](https://github.com/WordPress/gutenberg/blob/3130e85a217fc376102baaad18fdb6e07e1cb9c3/packages/block-editor/src/components/block-list/block.js#L34) could be improved, to clarify its relevance to the "block list" rendering. Related: https://github.com/WordPress/gutenberg/pull/21467#discussion_r405252631